### PR TITLE
Input Types Compatibility with OpenAI's API

### LINKED ISSUE #106: OpenAI compatible embedding API is not complete
### Feature request

https://platform.openai.com/docs/api-reference/embeddings/create
As mentioned in document, array of integers or array of arrays of integers could also be the input.

### Motivation

Integrity

### Your contribution

I'm willing to contribute, but not sure how to be compatible with current design.

### PR DESCRIPTION
**Pull request by nlaanait**
# What does this PR do?

Fixes #106 

More specifically, this PR brings TEI's `/embeddings` into full compatibility with all the input types supported by `openai`'s embeddings API ([v1 API reference](https://platform.openai.com/docs/api-reference/embeddings/create)).   
I attempted to add support for these additional types within the current design and with minimal changes to the existing logic. Main changes are:
1. Introduced `enum InputType` to support the different input data types: `String, u32, Vec<u32>`  (`router/src/http/types.rs`).
2. The existing `enum Input` is now strictly a container type (i.e. Single/Batch) using `InputType` as type.
3. Necessary methods to handle encoding of inputs in the form of int/array[int] were added as implementations to `EncodingInput` (`core/src/tokenization.rs`). 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: router/src/http/types.rs [hunk 1]
```diff
@@ -250,11 +250,36 @@ pub(crate) struct Rank {
...
+    SingleString(String),
+    SingleInt(u32),
+    VectorInt(Vec<u32>),
+}
+impl InputType {
+    pub(crate) fn count_chars(&self) -> usize {
+        match self {
+            InputType::SingleString(s) => s.chars().count(),
+            InputType::SingleInt(_) => 1,
+            InputType::VectorInt(v) => v.len(),
```
Comment (OlivierDehaene) [inline router/src/http/types.rs:265]: Is this how OpenAI counts when ids are given to the API? Or do they still count the chars by decoding the ids?

File: core/src/tokenization.rs [hunk 2]
```diff
@@ -125,12 +125,23 @@ fn encode_input(
...
+                "`inputs` must have less than {max_input_length} tokens. Given: {seq_len}"
+            )));
+        }
+        return inputs.try_into_encoding(position_offset);
+    }
 
     let inputs: EncodeInput = match inputs {
         EncodingInput::Single(s) => s.into(),
         EncodingInput::Dual(s1, s2) => (s1, s2).into(),
+        _ => Err(TextEmbeddingsError::Validation(
```
[the remark this answers is not in the document]
Comment (nlaanait) [reply core/src/tokenization.rs:141]: I'll give it a try.

File: router/src/http/types.rs (hunk 1 above)
Comment (nlaanait) [reply router/src/http/types.rs:265]: I'll look into it and modify this per my findings.

File: core/src/tokenization.rs [hunk 3]
```diff
@@ -125,12 +125,23 @@ fn encode_input(
         strategy: TruncationStrategy::LongestFirst,
         stride: 0,
     });
+    if inputs.is_encoded() {
```
[the remark this answers is not in the document]
Comment (nlaanait) [reply core/src/tokenization.rs]: Done

File: core/src/tokenization.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nlaanait) [reply core/src/tokenization.rs]: merged the logic above per your recommendation which made this branch irrelevant.

File: core/src/tokenization.rs [hunk 4]
```diff
@@ -163,13 +173,45 @@ pub struct Encoding {
...
+    fn try_into_encoding(&self, position_offset: usize) -> Result<Encoding, TextEmbeddingsError> {
+        match self {
+            EncodingInput::Vector(v) => Ok(Encoding {
+                input_ids: v.clone(),
+                token_type_ids: vec![0; v.len()],
+                position_ids: (position_offset as u32..(v.len() + position_offset) as u32)
+                    .collect::<Vec<_>>(),
+            }),
+            _ => Err(TextEmbeddingsError::Validation(
+                "`inputs` must be a vector of input_ids".to_string(),
```
[the remark this answers is not in the document]
Comment (nlaanait) [reply core/src/tokenization.rs]: I removed this.

File: core/src/tokenization.rs [hunk 5]
```diff
@@ -163,13 +173,45 @@ pub struct Encoding {
...
+            EncodingInput::Dual(s1, s2) => s1.len() + s2.len(),
+            EncodingInput::Vector(v) => v.len(),
+        }
+    }
+
+    fn try_into_encoding(&self, position_offset: usize) -> Result<Encoding, TextEmbeddingsError> {
+        match self {
+            EncodingInput::Vector(v) => Ok(Encoding {
+                input_ids: v.clone(),
+                token_type_ids: vec![0; v.len()],
```
[the remark this answers is not in the document]
Comment (nlaanait) [reply core/src/tokenization.rs]: I refactored in favor of building a `tokenizers::encoding` see: 
https://github.com/nlaanait/text-embeddings-inference/blob/2ee30448465239aefc6f212b18f9d51baac6b611/core/src/tokenization.rs#L143-L148

File: core/src/tokenization.rs [hunk 6]
```diff
@@ -163,13 +173,45 @@ pub struct Encoding {
...
+            EncodingInput::Single(s) => s.len(),
+            EncodingInput::Dual(s1, s2) => s1.len() + s2.len(),
+            EncodingInput::Vector(v) => v.len(),
+        }
+    }
+
+    fn try_into_encoding(&self, position_offset: usize) -> Result<Encoding, TextEmbeddingsError> {
+        match self {
+            EncodingInput::Vector(v) => Ok(Encoding {
+                input_ids: v.clone(),
```
[the remark this answers is not in the document]
Comment (nlaanait) [reply core/src/tokenization.rs]: the validation is now performed via a call to `tokenizer.decode` see:
https://github.com/nlaanait/text-embeddings-inference/blob/2ee30448465239aefc6f212b18f9d51baac6b611/core/src/tokenization.rs#L168-L195

File: router/src/http/types.rs (hunk 1 above)
Comment (nlaanait) [reply router/src/http/types.rs:265]: @OlivierDehaene I looked through the source of `openai-python` v1.7.1 and couldn't find a reference to  characters counting in the embeddings API.
Should `count_chars` return `0` for `InputType::SingleInt, InputType::VectorInt` for correctness?
LMK how you want to proceed.

**Comment by matatonic** [comment]
Is it possible to also add the correct routing "/v1/embeddings" as per the API? https://platform.openai.com/docs/api-reference/embeddings. Without this change it's not possible to use the environment variable OPENAI_BASE_URL in a consistent way with this service. Ex:
# For text-embeddings-inference:
OPENAI_BASE_URL=http://text-embeddings-inference:80
# Normally:
OPENAI_BASE_URL=http://<server ip>:<port>/v1

Also, for reference, the text-generation-webui has a more compatible implementation, including base64 json encoding (undocumented, but in the openai python package):
https://github.com/oobabooga/text-generation-webui/tree/main/extensions/openai

**Pull request merged**